packages/core/src/modules/message-pìckup/MessagePickupApi.ts

```diff
@@ -66,13 +77,52 @@ export class MessagePickupApi<MPPs extends MessagePickupProtocol[] = [V1MessageP
    */
   public async queueMessage(options: QueueMessageOptions): Promise<QueueMessageReturnType> {
     this.logger.debug('Queuing message...')
-    const connectionRecord = await this.connectionService.getById(this.agentContext, options.connectionId)
+    const { connectionId, message, recipientKey } = options
+    const connectionRecord = await this.connectionService.getById(this.agentContext, connectionId)
 
-    const messageRepository = this.agentContext.dependencyManager.resolve<MessageRepository>(
-      InjectionSymbols.MessageRepository
+    const messagePickupRepository = this.agentContext.dependencyManager.resolve<MessagePickupRepository>(
+      InjectionSymbols.MessagePickupRepository
     )
 
-    await messageRepository.add(connectionRecord.id, options.message)
+    await messagePickupRepository.addMessage({ connectionId: connectionRecord.id, recipientKey, payload: message })
+  }
+
+  /**
+   * Deliver messages in the Message Pickup Queue for a given connection and key (if specified).
+   *
+   * Note that this is only available when there is an active session with the recipient. Message
+   * Pickup protocol messages themselves are not added to pickup queue
+   *
+   */
+  public async deliverQueuedMessages(options: DeliverQueuedMessagesOptions) {
```

Comment:
Should this take a sessionId instead, so you wont have the case where there's no session? Maybe that'll complicate it without benefits

I think this should at least return an object, where it e.g. optionally returns the message, and a state which indicates how the delivery went (e.g. for now no open session, or delivered). Also will this deliver all messages? Or should you call this in a loop?

Reply:
In order to make it more flexible, I've added an API method to get live sessions and make this one to use sessionId instead. This way, the only possibility is that it will proceed with the delivery (e.g. the delivery message has been properly submitted) or an error happened (MessageSendingError or so), so it will throw and can be handled by the caller.

This method, now called `deliverMessagesFromQueue`, will retrieve any pending messages from the queue and send the initial batch of `batchSize` messages, starting a pickup loop that will be finished once all messages are delivered (the client will automatically request more messages when the status shows that there are pending messages). This initial `batchSize` number is sent as an input parameter or taken from configuration.

There is also another method called `deliverMessages` which is similar but intended to send specific messages. It will not retrieve other messages from the queue. This can be more appropriate when several messages for a given mediatee are received within a short time frame, as there will not be need of querying the queue multiple times.